aws-codeguruprofiler-profilinggroup/src/main/java/software/amazon/codeguruprofiler/profilinggroup/AgentPermissionHelper.java

```diff
@@ -0,0 +1,46 @@
+package software.amazon.codeguruprofiler.profilinggroup;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import software.amazon.awssdk.services.codeguruprofiler.CodeGuruProfilerClient;
+import software.amazon.awssdk.services.codeguruprofiler.model.CodeGuruProfilerException;
+import software.amazon.awssdk.services.codeguruprofiler.model.GetPolicyRequest;
+import software.amazon.awssdk.services.codeguruprofiler.model.GetPolicyResponse;
+import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
+
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+
+public class AgentPermissionHelper {
+    private static final CodeGuruProfilerClient profilerClient = CodeGuruProfilerClientBuilder.create();
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    @FunctionalInterface
+    public interface GetPrincipalsFunction<Proxy, PgName, Principals> {
+        Principals apply(Proxy s, PgName t);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static List<String> getPrincipalsFromPolicy(AmazonWebServicesClientProxy proxy, String pgName) {
+        GetPolicyRequest getPolicyRequest = GetPolicyRequest.builder().profilingGroupName(pgName).build();
+        GetPolicyResponse getPolicyResponse = proxy.injectCredentialsAndInvokeV2(getPolicyRequest, profilerClient::getPolicy);
+        String policyInJson = getPolicyResponse.policy();
+
+        if (policyInJson == null) return emptyList();
+
+        try {
+            Map<String, List<Map<String, Map<String, Object>>>> policyMap = objectMapper.readValue(policyInJson, Map.class);
+            Object principals = policyMap.get("Statement").get(0).get("Principal").get("AWS");
+
+            if (principals instanceof String) {
+                return singletonList((String) principals);
+            } else {
+                return (List<String>) principals;
+            }
+        } catch (Exception e) {
+            throw CodeGuruProfilerException.create("Failed to deserialize policy: " + policyInJson, e);
```

Comment:
This exception is from the CodeGuru Profiler Java SDK itself, so it's a bit weird to use here -- it makes it seem as this is an error that came from a service call. I'm not even sure how the top-level handler will show it...?

 Should we perhaps use an `CfnServiceInternalErrorException` or something like that?

Reply:
The reason I wrap this with our service exception is because I felt, if the policyInJson cannot be deserialized here properly, it should be an error from our service response. The response should be standardised for 

1. null if no policy attached
2. standard json in the recognisable format that we can deserialize here. 

Any of the two assumption violated should be an issue from our Service instead of cfn error. ><. I only expect exception caused by violating the assumption above here.

Reply:
@ivoanjo is right on this one :) All exceptions that are returned to CloudFormation (who is the caller of the handler) should be wrapped in the `CfnXXXException` variations in order to know how to handle them on their side, e.g. retrying, rolling back, etc.

See https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/issues/243 along with https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-contract-errors.html and https://www.alexdebrie.com/posts/cloudformation-custom-resources/#tips-and-tricks-for-writing-custom-resources.

Reply:
⬆️ what that wise man said